crates/store/src/state.rs

```diff
@@ -783,61 +751,110 @@ impl State {
         account_ids: &[AccountId],
         nullifiers: &[Nullifier],
         unauthenticated_notes: BTreeSet<NoteId>,
+        reference_blocks: BTreeSet<BlockNumber>,
     ) -> Result<BlockInputs, GetBlockInputsError> {
+        // Get the note inclusion proofs from the DB.
+        // We do this first so we have to acquire the lock to the state just once. There we need the
+        // reference blocks of the note proofs to get their authentication paths in the chain MMR.
+        let unauthenticated_note_proofs = self
+            .db
+            .select_note_inclusion_proofs(unauthenticated_notes)
+            .await
+            .map_err(GetBlockInputsError::SelectNoteInclusionProofError)?;
```

Comment:
This selects notes from the DB by note ID, but the notes table does not have the note id as a primary key, so this has to scan the entire table, if I'm not mistaken.
I also wonder if we need batch index and note index as primary keys.

Reply:
That's correct; and we are missing quite a lot of indices. I think we only have like three indices explicitly created - so almost any non-primary/composite key lookup will be a scan at the moment.

There is #712 where we intend to go through these, maybe even figure out a way to test for lack of `scan` in the query plans.

I've also been eyeing `diesel` because I'm curious, and having compile-time checked queries would reduce the need for a lot of tests that only serve to check the sql statements.